hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/multipart/MultipartDownloader.java

```diff
@@ -0,0 +1,103 @@
+package org.apache.hadoop.fs.s3a.multipart;
+
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.List;
+
+public final class MultipartDownloader {
+    private static final Log LOG = LogFactory.getLog(MultipartDownloader.class);
+
+    private final long partSize;
+    private final ListeningExecutorService downloadExecutorService;
+    private final PartDownloader partDownloader;
+    private final long chunkSize;
+    private final long bufferSize;
+
+    public MultipartDownloader(long partSize, ListeningExecutorService downloadExecutorService, PartDownloader partDownloader, long chunkSize, long bufferSize) {
+        this.partSize = partSize;
+        this.downloadExecutorService = downloadExecutorService;
+        this.partDownloader = partDownloader;
+        this.chunkSize = chunkSize;
+        this.bufferSize = bufferSize;
+    }
+
+    public AbortableInputStream download(final String bucket, final String key, long rangeStart, long rangeEnd) {
+        final long size = rangeEnd - rangeStart;
+        int numParts = (int) Math.ceil((double) size / partSize);
+
+        final OrderingQueue orderingQueue = new OrderingQueue(rangeStart, size, bufferSize);
+
+        final List<ListenableFuture<?>> partFutures = Lists.newArrayList();
+
+        for (long i = 0; i < numParts; i++) {
+            final long partRangeStart = rangeStart + i * partSize;
+            final long partRangeEnd = i == numParts - 1 ? rangeEnd : partRangeStart + partSize;
+
+            ListenableFuture<?> partFuture = downloadExecutorService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    LOG.info(String.format("Downloading part %d - %d", partRangeStart, partRangeEnd));
```

Comment:
Would really try to avoid `String.format` to skip format` parsing and string generation overhead